### DISCUSSION TITLE
Help tracing nested resolvers

### DISCUSSION THREAD

**Discussion by 2snEM6** [hayes/pothos#1066] [Q&A] [answered]
`@pothos/tracing-sentry` seems to have a hard time setting up tracing when nested resolvers are involved.

Let's say I have the following query:

```graphql
query TestQuery {
  info {
    id
    name
  }
}
```

And Pothos builder is configured to trace all resolvers:

```javascript
tracing: {
    default: (config) => true,
    wrap: (resolver, options) => traceResolver(resolver, options),
},
```

And `traceResolver` is setup this way:

```typescript
const traceResolver = createSentryWrapper({
  includeArgs: true,
  includeSource: true,
});
```

Each span created on each resolver execution (id, name) should be nested under `info`, but instead they're not nested and `id` and `name` spans are at the same flat level as their parent `info` under the transaction root.

I modified `createSentryWrapper` to make use of `Sentry.runWithAsyncContext` that uses AsyncLocalStorage to prevent context leaks across async executions. This partially fixes nesting for `DB` and outgoing `http` spans created automatically by Sentry instrumentation. However nested resolver spans remain flat. See:

See:

<img width="858" alt="image" src="https://github.com/hayes/pothos/assets/7485885/3b6bd995-79a1-4f49-9b02-2162b79b9835">

This is my updated implementation of `createSentryWrapper`:

```typescript
export function createSentryWrapper<T = unknown>(
  options?: SentryWrapperOptions<T>,
) {
  return <Context extends object = object>(
      resolver: GraphQLFieldResolver<unknown, Context, Record<string, unknown>>,
      fieldOptions: T,
      tracingOptions?: SentryWrapperOptions<T>,
    ): GraphQLFieldResolver<unknown, Context, Record<string, unknown>> =>
    (source, args, ctx, info) => {
      return Sentry.runWithAsyncContext(
        () => {
          const scope = Sentry.getCurrentHub().getScope();
          const transaction = scope.getTransaction();

          if (!transaction) {
            return resolver(source, args, ctx, info);
          }
          // 1. Identify the parent span:
          // This checks the context if there's an active span for the current GraphQL request.
          const parentSpan = scope.getSpan() || transaction;

          const tags = {
            [AttributeNames.FIELD_NAME]: info.fieldName,
            [AttributeNames.FIELD_PATH]: pathToString(info),
            [AttributeNames.FIELD_TYPE]: info.returnType.toString(),
          };

          const data: Record<string, unknown> = {};

          if (tracingOptions?.includeArgs ?? options?.includeArgs) {
            data[AttributeNames.FIELD_ARGS] = args;
          }

          if (tracingOptions?.includeSource ?? options?.includeSource) {
            if (info.fieldNodes?.[0]) {
              data[AttributeNames.SOURCE] = print(info.fieldNodes[0]);
            }
          }
          const span = parentSpan.startChild({
            op: 'graphql.resolve',
            description: info.fieldName,
            tags,
            data,
          }) as Span;

          scope?.setSpan(span);

          tracingOptions?.onSpan?.(span, fieldOptions, source, args, ctx, info);
          options?.onSpan?.(span, fieldOptions, source, args, ctx, info);

          return runFunction(
            () => resolver(source, args, ctx, info),
            () => {
              scope?.setSpan(parentSpan);
              span.finish();
            },
          );
        },
        {
          reuseExisting: false,
        },
      );
    };
}
```

I am using `graphql-yoga` and the Envelope Sentry plugin which starts a transaction on each GraphQL execution.

Thank you

**Comment by hayes** [comment]
The tracing plugin has helper to manage connecting parent spans.  You can see example usage in the open telemetry plugin:  https://github.com/hayes/pothos/blob/main/packages/tracing-opentelemetry/src/index.ts#L32

**Reply by 2snEM6** [reply, marked as answer]
Thank you! I ended up doing something like this

```typescript
export function createSentryWrapper<T = unknown>(
  options?: SentryWrapperOptions<T>,
) {
  return <Context extends object = object>(
      resolver: GraphQLFieldResolver<unknown, Context, Record<string, unknown>>,
      fieldOptions: T,
      tracingOptions?: SentryWrapperOptions<T>,
    ): GraphQLFieldResolver<unknown, Context, Record<string, unknown>> =>
    (source, args, ctx, info) => {
      return Sentry.runWithAsyncContext(
        () => {
          const scope = Sentry.getCurrentHub().getScope();
          const transaction = scope.getTransaction();

          if (!transaction) {
            return resolver(source, args, ctx, info);
          }

          const span = createSpanWithParent<Sentry.Span>(
            ctx,
            info,
            (path, parentSpan) => {
              const computedParentSpan =
                parentSpan || scope.getSpan() || transaction;

              const tags = {
                [AttributeNames.FIELD_NAME]: info.fieldName,
                [AttributeNames.FIELD_PATH]: path,
                [AttributeNames.FIELD_TYPE]: info.returnType.toString(),
              };

              const data: Record<string, unknown> = {};

              if (tracingOptions?.includeArgs ?? options?.includeArgs) {
                data[AttributeNames.FIELD_ARGS] = args;
              }

              if (tracingOptions?.includeSource ?? options?.includeSource) {
                if (info.fieldNodes?.[0]) {
                  data[AttributeNames.SOURCE] = print(info.fieldNodes[0]);
                }
              }

              const childSpan = computedParentSpan.startChild({
                op: 'graphql.resolve',
                description: info.fieldName,
                tags,
                data,
              });

              scope?.setSpan(childSpan);

              return childSpan;
            },
          );

          tracingOptions?.onSpan?.(span, fieldOptions, source, args, ctx, info);
          options?.onSpan?.(span, fieldOptions, source, args, ctx, info);

          return runFunction(
            () => resolver(source, args, ctx, info),
            (err: unknown) => {
              if (err && !isOriginalGraphQLError(err)) {
                if (
                  (err instanceof ApiError && !err.isUserError) ||
                  !(err instanceof ApiError)
                ) {
                  const code = extractErrorCode(err as unknown);

                  const errorInfo = {
                    code,
                    operationName: info.operation.name?.value,
                    query: info.operation.loc?.source.body,
                    variableValues: info.variableValues,
                  };

                  Sentry.withScope((errScope) => {
                    attachInfoToScope(errScope, errorInfo);
                    Sentry.captureException(err);
                  });
                }
              }
              span.finish();
            },
          );
        },
        {
          reuseExisting: false,
        },
      );
    };
}
```

Not sure if this bit: `const computedParentSpan =
                parentSpan || scope.getSpan() || transaction;` is accurate, but it seems to work!

**Reply by 2snEM6** [reply]
Something I noticed is that it's better to capture errors here than in the operation root resolver, this way Sentry highlights which resolver is faulty. So, I ended up moving the error capturing logic from the graphql-yoga setup logic to here. I wrap all resolvers

**Reply by hayes** [reply]
makes sense.   I would be careful about wrapping all resolvers if you are sensitive to performance, this will add a lot of overhead to your requests.   I think for a lot of applications this doesn't really matter, but be aware this can get expensive under heavy loads.

**Reply by 2snEM6** [reply]
Thank you! On our case we measured performance and it didn't really make any difference, but I will keep this in mind for the future. Before this I wanted to only wrap fields that had resolver functions defined explicitly, but for whatever reason even if i don't define the resolver function for a field, it's still there. Do you know why?

**Reply by 2snEM6** [reply]
In this piece of code basically:

<img width="216" alt="image" src="https://github.com/hayes/pothos/assets/7485885/f0ed5d0d-d157-44ef-a588-a478113225f2">

**Reply by hayes** [reply]
It makes plugins a lot harder to write if the resolve isn't guaranteed to be a function.   GraphQL will use defaultFieldResolver from the graphql package if no resolver is defined.   Pothos will add this as a default resolver if none is provided.   After plugins are applied, we check of the resolver is defaultFieldResolver, and then remove it if it hasn't been wrapped.  You can probably also check for that to see if you should wrap.   

This util is built for that purpose https://github.com/hayes/pothos/blob/main/packages/plugin-tracing/src/util.ts#L25. And also handles the expose fields which do have resolvers, but don't do anything interesting

**Reply by 2snEM6** [reply]
This is great! I will take a look, thank you for all your help.

Cheers!
